File: Problem18_ContiguousArray.java
// Time Complexity : O(n) -> n: length of the array
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int findMaxLength(int[] nums) {
        if(nums == null || nums.length == 0)
            return 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        int rSum = 0, max = 0;
        map.put(0, -1);
        for(int i = 0; i < nums.length; i++) {
            rSum += (nums[i] == 1 ? 1 : -1);
            if(map.containsKey(rSum)) {
                max = Math.max(max, i - map.get(rSum));
            } else {
                map.put(rSum, i);
            }
        }
        return max;
    }
}

File: Problem 18 Contiguous Array.py
# Did this code successfully run on Leetcode: Yes
# Time Complexity: O(n)
# Space Complexity: O(n) since mapper takes sum & sum can have n values.
# Logic: treat 0 as -1 & keep checking sum to be equal to 0. Store sum with index in a dic for cal max length.

class Solution:
    def findMaxLength(self, nums):
        mapper = {0: -1}
        sum = 0
        maxlen = 0
        for i in range(len(nums)):
            if nums[i] == 0:
                sum += -1
            else:
                sum += 1
            if sum in mapper:
                maxlen = max(maxlen, i - mapper[sum])
            else:
                mapper[sum] = i
        return maxlen
